data/generate_embeddings.py

```diff
@@ -0,0 +1,598 @@
+        overlap_ratio=overlap_ratio
+    )
+    
+    chunks: List[str] = [detokenize_tokens(tc, encoder) for tc in token_chunks]
+    return chunks
+
+def process_file(
+    provider: EmbeddingProvider,
+    input_path: Path,
+    output_dir: Path,
+    encoder,
+    max_tokens: int,
+    overlap_ratio: float,
+    batch_size: int,
+    path_to_city: Dict[str, str],
+    basename_to_city: Dict[str, str],
+    polite_delay: float = DEFAULT_POLITE_DELAY,
+    retry_attempts: int = DEFAULT_RETRY_ATTEMPTS,
+    retry_min_wait: int = DEFAULT_RETRY_MIN_WAIT,
+    retry_max_wait: int = DEFAULT_RETRY_MAX_WAIT,
+) -> Tuple[int, Optional[int]]:
+    """Process a single input file and generate embeddings."""
+    print(f"Processing: {input_path.name}")
+    
+    raw_text = read_file_content(input_path)
+    if not raw_text:
+        print(f"  Warning: No content found in {input_path.name}")
+        return 0, None
+
+    chunks = build_chunks(
+        raw_text, 
+        max_tokens=max_tokens, 
+        overlap_ratio=overlap_ratio, 
+        encoder=encoder
+    )
+    
+    if not chunks:
+        print(f"  Warning: No chunks generated from {input_path.name}")
+        return 0, None
+        
+    city = infer_city_from_metadata(input_path, path_to_city, basename_to_city)
+    start_index = get_next_global_chunk_id(output_dir)
+
+    # Process in batches
+    total_written = 0
+    for i in range(0, len(chunks), batch_size):
+        batch = chunks[i : i + batch_size]
+        
+        try:
+            vectors = provider.embed_batch(
+                batch,
+                retry_attempts=retry_attempts,
+                retry_min_wait=retry_min_wait,
+                retry_max_wait=retry_max_wait
+            )
+            
+            # Validate embeddings
+            if not validate_embeddings(vectors, len(batch)):
+                print(f"  Error: Validation failed for batch starting at chunk {i}")
+                continue
+            
+            # Save chunks
+            for j, (text, vec) in enumerate(zip(batch, vectors)):
+                save_chunk_json(
+                    output_dir=output_dir,
+                    city=city,
+                    chunk_id=start_index + total_written + j,
+                    text=text,
+                    embedding=vec,
+                    source_file=input_path.name,
+                    model=provider.model,
+                )
+            
+            total_written += len(batch)
+            print(f"  Processed batch {i//batch_size + 1}: {len(batch)} chunks")
+            
+            # Polite pacing
+            time.sleep(polite_delay)
+            
+        except Exception as e:
+            print(f"  Error processing batch {i//batch_size + 1}: {e}")
+            continue
+
+    last_index = start_index + total_written - 1 if total_written > 0 else None
+    print(f"  Completed: {total_written} chunks written for city '{city}'")
+    
+    return total_written, last_index
```

Comment:
Track and log the total time taken for each file and batch for performance analysis.
If batch API call fails, optionally retry at the batch level before skipping.